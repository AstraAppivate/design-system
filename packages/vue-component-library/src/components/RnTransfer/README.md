---
title: Transfer
description: A component which allows you to transfer items from one panel to another
---

# Overview

Transfers items from one panel to another 

## Usage

The transfer component can be used anywhere within an HTML form.

#### Properties

| Name         | Type                  | Required | Description                                                                                                |
| ------------ | --------------------- | -------- | ---------------------------------------------------------------------------------------------------------- |
| listData     | Array                 | True     | Provides an array of data to be used within the transfer component (see example below)                     |
| leftHeader   | String                | False    | Provides a header for the left panel                                                                       |
| rightHeader  | String                | False    | Provides a header for the right panel                                                                      |

#### Example

![Toggle example](images/transfer.gif)

```js
const listData = [
        {
          "id": "stfaith",
          "location": 'St Faith', 
          "list" : 'left',
        },
        {
          "id": "portsmouth",
          "location": 'Portsmouth', 
          "list" : 'right',
        },
        {
          "id": "gosport",
          "location": 'Gosport', 
          "list" : 'left',
        },
        {
          "id": "southampton",
          "location": 'Southampton', 
          "list" : 'left',
        },
      ]

<rn-transfer :listData="listData"></rn-transfer>
```
