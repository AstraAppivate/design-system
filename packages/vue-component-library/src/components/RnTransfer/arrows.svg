<svg xmlns="http://www.w3.org/2000/svg" width="12" height="10" viewBox="0 0 12 10">
  <g fill="#3E5667" fill-rule="evenodd" transform="translate(0 .199)">
    <polygon points="10.548 9.095 6 4.548 10.548 0 11.969 1.421 8.842 4.548 11.969 7.674" transform="rotate(180 8.984 4.548)"/>
    <polygon points="4.548 9.095 0 4.548 4.548 0 5.969 1.421 2.842 4.548 5.969 7.674" transform="rotate(180 2.984 4.548)"/>
  </g>
</svg>
