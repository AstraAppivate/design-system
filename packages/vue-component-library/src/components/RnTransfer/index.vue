<template lang="html">
  <div class="rn-transfer">
    <div class="column-left"> 
      <h1 :if="leftHeader">{{leftHeader}}</h1>
      <ul aria-label="checkbox-group">
        <li v-for="(item, index) in leftList" :key="`item-${index}`">
          <label>
            <input type="checkbox" />
            <span>{{item.location}}</span>
          </label>
        </li>
      </ul>
    </div>
    <div class="transfer-buttons">
      <span class="left" @click="to_left">&lt;</span>
      <span class="right" @click="to_right">&gt;</span>
    </div>
    <div class="column-right">
      <h1 :if="rightHeader">{{rightHeader}}</h1>
      <ul aria-label="checkbox-group">
        <li v-for="(item, index) in rightList" :key="`item-${index}`">
          <label>
            <input type="checkbox" />
            <span>{{item.location}}</span>
          </label>
        </li>
      </ul>
    </div>
  </div>
</template>

<script src="./scripts.js"></script>
