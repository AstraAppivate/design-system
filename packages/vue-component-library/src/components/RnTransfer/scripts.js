import '@royalnavy/css-framework/src/components/_transfer.scss'

export default {
  name: 'RnTransfer',

  props: {
    listData: !Array,
    leftHeader: String,
    rightHeader: String,
  },

  data: () => ({
    listDataMutable: [],
  }),
  created() {
    this.listDataMutable = this.listData
  },

  computed: {
    leftList() {
      return this.listDataMutable.filter(item => item.list === 'left')
    },

    rightList() {
      return this.listDataMutable.filter(item => item.list === 'right')
    },
  },

  methods: {
    move(item, side) {
      const itemCopy = item

      if (itemCopy.checked) itemCopy.list = side
      itemCopy.checked = false
      this.$emit('moved', this.listData)

      return itemCopy
    },

    to_left() {
      this.rightList.map(item => this.move(item, 'left'))
    },

    to_right() {
      this.leftList.map(item => this.move(item, 'right'))
    },
  },
}
