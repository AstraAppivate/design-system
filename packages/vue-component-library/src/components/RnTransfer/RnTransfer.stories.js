import { storiesOf } from '@storybook/vue'

import RnTransfer from './index.vue'

storiesOf('Transfer', module)
  .add('Basic', () => ({
    components: { RnTransfer },
    data() {
      return {
        listData: [
          {
            "id": "stfaith",
            "location": 'St Faith', 
            "list" : 'left',
            "checked" : false,
          },
          {
            "id": "portsmouth",
            "location": 'Portsmouth', 
            "list" : 'right',
            "checked" : false,
          },
          {
            "id": "gosport",
            "location": 'Gosport', 
            "list" : 'left',
            "checked" : false,
          },
          {
            "id": "southampton",
            "location": 'Southampton', 
            "list" : 'left',
            "checked" : false,
          },
        ]
      }
    },
    template: `
      <rn-transfer :listData="listData" style="max-width: 600px;"></rn-transfer>
    `,
  }))
  .add('With Headers', () => ({
    components: { RnTransfer },
    data() {
      return {
        listData: [
          {
            "id": "stfaith",
            "location": 'St Faith', 
            "list" : 'left',
            "checked" : false,
          },
          {
            "id": "portsmouth",
            "location": 'Portsmouth', 
            "list" : 'right',
            "checked" : false,
          },
          {
            "id": "gosport",
            "location": 'Gosport', 
            "list" : 'left',
            "checked" : false,
          },
          {
            "id": "southampton",
            "location": 'Southampton', 
            "list" : 'left',
            "checked" : false,
          },
        ]
      }
    },
    template: `
      <rn-transfer :listData="listData" leftHeader="Left header" rightHeader="Right header" style="max-width: 600px;"></rn-transfer>
    `,
  }))
