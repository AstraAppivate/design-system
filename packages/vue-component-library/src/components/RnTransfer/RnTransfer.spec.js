import { shallow } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import RnTransfer from '@/src/components/RnTransfer/index.vue'

describe('RnTransfer.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(RnTransfer, {
      propsData: {
        leftHeader: 'This is a test title',
        rightHeader: 'This is a test title',
        listData: [
          {
            "id": "stfaith",
            "location": 'St Faith', 
            "list" : 'left',
            "checked" : false,
          },
          {
            "id": "portsmouth",
            "location": 'Portsmouth', 
            "list" : 'right',
            "checked" : false,
          },
          {
            "id": "gosport",
            "location": 'Gosport', 
            "list" : 'left',
            "checked" : false,
          },
          {
            "id": "southampton",
            "location": 'Southampton', 
            "list" : 'left',
            "checked" : false,
          },
        ]
      },
    })
  })

  it('snapshot: has same HTML structure', () => {
    const renderer = createRenderer()

    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) {
        throw new Error(err)
      }

      expect(str).toMatchSnapshot()
    })
  })
  it('Displays a left Header', () => {
    expect(wrapper.find('.column-left > h1').text()).toBe('This is a test title')
  })
  it('Displays a right Header', () => {
    expect(wrapper.find('.column-right > h1').text()).toBe('This is a test title')
  })
  it('Has 3 items in the left column', () => {
    expect(wrapper.find('.column-left > ul').findAll('li').length).toBe(3)
  })
  it('Has 1 item in the right column', () => {
    expect(wrapper.find('.column-right > ul').findAll('li').length).toBe(1)
  })
  it('Moves items from the right list to the left list', () => {
    wrapper.find('.column-right > ul > li > label > input[name="portsmouth"]').setChecked()
    wrapper.find('.transfer-buttons > .left').trigger('click')
    expect(wrapper.emitted().moved).toBeTruthy()
  })
})


