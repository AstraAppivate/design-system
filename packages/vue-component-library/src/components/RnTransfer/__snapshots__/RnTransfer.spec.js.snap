// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RnTransfer.vue snapshot: has same HTML structure 1`] = `
<div class="rn-transfer">
  <div class="column-left">
    <h1>This is a test title</h1>
    <ul aria-label="checkbox-group">
      <li><label><input type="checkbox" name="stfaith" value="St Faith"> <span>St Faith</span></label></li>
      <li><label><input type="checkbox" name="gosport" value="Gosport"> <span>Gosport</span></label></li>
      <li><label><input type="checkbox" name="southampton" value="Southampton"> <span>Southampton</span></label></li>
    </ul>
  </div>
  <div class="transfer-buttons"><span class="left"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="10" viewBox="0 0 12 10"><g fill="#fff" fill-rule="evenodd" transform="translate(0 .199)"><polygon points="10.548 9.095 6 4.548 10.548 0 11.969 1.421 8.842 4.548 11.969 7.674" transform="rotate(0 8.984 4.548)"></polygon> <polygon points="4.548 9.095 0 4.548 4.548 0 5.969 1.421 2.842 4.548 5.969 7.674" transform="rotate(0 2.984 4.548)"></polygon></g></svg></span> <span class="right"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="10" viewBox="0 0 12 10"><g fill="#fff" fill-rule="evenodd" transform="translate(0 .199)"><polygon points="10.548 9.095 6 4.548 10.548 0 11.969 1.421 8.842 4.548 11.969 7.674" transform="rotate(180 8.984 4.548)"></polygon> <polygon points="4.548 9.095 0 4.548 4.548 0 5.969 1.421 2.842 4.548 5.969 7.674" transform="rotate(180 2.984 4.548)"></polygon></g></svg></span></div>
  <div class="column-right">
    <h1>This is a test title</h1>
    <ul aria-label="checkbox-group">
      <li><label><input type="checkbox" name="portsmouth" value="Portsmouth"> <span>Portsmouth</span></label></li>
    </ul>
  </div>
</div>
`;
